### PR TITLE
docs(graphql): document fieldResolverEnhancers

### PR DESCRIPTION
**Pull request by gempain**
## PR Checklist

- [x] The commit message follows our guidelines: https://github.com/nestjs/docs.nestjs.com/blob/master/CONTRIBUTING.md

## PR Type

```
[x] Other... Please describe: docs
```

## What is the current behavior?

Issue Number: https://github.com/nestjs/graphql/issues/760

## Does this PR introduce a breaking change?
```
[x] No
```

### CODE REVIEW

**Comment by kamilmysliwiec** [comment]
Thanks @gempain! @johnbiundo could you review? API-wise looks good :)

File: content/graphql/guards-interceptors.md [hunk 1]
```diff
@@ -85,3 +85,27 @@ async upvotePost(
...
+GraphQLModule.forRoot({
+  fieldResolverEnhancers: ['interceptors']
+}),
+```
+
+> **Warning** Enabling interceptors for field resolvers can cause performance issues when you are returning lots of records and your field resolver is executed thousands of times. For this reason, when you enable `fieldResolverEnhancers`, we advise you to skip execution of interceptors that are not strictly necessary for your field resolvers. You can do this using the following helper method:
+
+```typescript
+export function isResolvingGraphQLField(context: ExecutionContext): boolean {
+  if (context.getType().toString() === 'graphql') {
```
Comment (kamilmysliwiec) [inline content/graphql/guards-interceptors.md:103]: One little suggestion: you can use the `GqlContextType` as stated here https://docs.nestjs.com/fundamentals/execution-context#current-application-context

```typescript
if (context.getType<GqlContextType>() === 'graphql') {
```

File: content/graphql/guards-interceptors.md [hunk 2]
```diff
@@ -85,3 +85,27 @@ async upvotePost(
 ```
 
 > info **Hint** In the above example, we have assumed that the `user` object is assigned to the context of your GraphQL application.
+
+#### Interceptors at the `@ResoleField()` method level
```
Comment (johnbiundo) [inline content/graphql/guards-interceptors.md:89]:
```suggestion
#### Interceptors at the `@ResolveField()` method level
```

File: content/graphql/guards-interceptors.md [hunk 3]
```diff
@@ -85,3 +85,27 @@ async upvotePost(
...
+
+In a GraphQL context, interceptors [do not access the full GraphQL response](https://github.com/nestjs/graphql/issues/320#issuecomment-511193229): they only get the response of the `@Query()`/`@Mutation()` method, but not the complete response with the resolved fields. You can tell Nest to execute interceptors (as well as guards and filters) for methods annotated `@ResolveField()` by setting the `fieldResolverEnhancers` option in `GqlModuleOptions`:
+
+```typescript
+GraphQLModule.forRoot({
+  fieldResolverEnhancers: ['interceptors']
+}),
+```
+
+> **Warning** Enabling interceptors for field resolvers can cause performance issues when you are returning lots of records and your field resolver is executed thousands of times. For this reason, when you enable `fieldResolverEnhancers`, we advise you to skip execution of interceptors that are not strictly necessary for your field resolvers. You can do this using the following helper method:
```
Comment (johnbiundo) [inline content/graphql/guards-interceptors.md:99]:
```suggestion
> **Warning** Enabling interceptors for field resolvers can cause performance issues when you are returning lots of records and your field resolver is executed thousands of times. For this reason, when you enable `fieldResolverEnhancers`, we advise you to skip execution of interceptors that are not strictly necessary for your field resolvers. You can do this using the following helper function:
```

File: content/graphql/guards-interceptors.md [hunk 4]
```diff
@@ -96,11 +96,11 @@ GraphQLModule.forRoot({
 }),
 ```
 
-> **Warning** Enabling interceptors for field resolvers can cause performance issues when you are returning lots of records and your field resolver is executed thousands of times. For this reason, when you enable `fieldResolverEnhancers`, we advise you to skip execution of interceptors that are not strictly necessary for your field resolvers. You can do this using the following helper method:
+> **Warning** Enabling interceptors for field resolvers can cause performance issues when you are returning lots of records and your field resolver is executed thousands of times. For this reason, when you enable `fieldResolverEnhancers`, we advise you to skip execution of interceptors that are not strictly necessary for your field resolvers. You can do this using the following helper function:
 
 ```typescript
 export function isResolvingGraphQLField(context: ExecutionContext): boolean {
-  if (context.getType().toString() === 'graphql') {
+  if (context.getType<GqlContextType>().toString() === 'graphql') {
```
Comment (johnbiundo) [inline content/graphql/guards-interceptors.md:103]: I think you can simplify further (see http://localhost:4200/fundamentals/execution-context#current-application-context)

```suggestion
  if (context.getType<GqlContextType>() === 'graphql') {
```

Comment (gempain) [reply content/graphql/guards-interceptors.md:103]: Had missed that from @kamilmysliwiec's comment.

**Comment by kamilmysliwiec** [comment]
Oh, I've missed one thing! Note that with the `fieldResolverEnhancers` you can enable interceptors, guards, and filters (not only interceptors). Example:

```typescript
fieldResolverEnhancers: ['interceptors', 'guards']
```

File: content/graphql/guards-interceptors.md [hunk 5]
```diff
@@ -85,3 +85,27 @@ async upvotePost(
 ```
 
 > info **Hint** In the above example, we have assumed that the `user` object is assigned to the context of your GraphQL application.
+
+#### Interceptors at the `@ResolveField()` method level
```
Comment (johnbiundo) [inline content/graphql/guards-interceptors.md:89]:
```suggestion
#### Execute enhancers at the field/method level
```

File: content/graphql/guards-interceptors.md [hunk 6]
```diff
@@ -85,3 +85,27 @@ async upvotePost(
...
+
+In a GraphQL context, interceptors [do not access the full GraphQL response](https://github.com/nestjs/graphql/issues/320#issuecomment-511193229): they only get the response of the `@Query()`/`@Mutation()` method, but not the complete response with the resolved fields. You can tell Nest to execute interceptors (as well as guards and filters) for methods annotated `@ResolveField()` by setting the `fieldResolverEnhancers` option in `GqlModuleOptions`:
```
Comment (johnbiundo) [inline content/graphql/guards-interceptors.md:91]:
```suggestion
In the GraphQL context, Nest does not run **enhancers** (the generic name for interceptors, guards and filters) at the field level [see this issue](https://github.com/nestjs/graphql/issues/320#issuecomment-511193229): they only run for the top level `@Query()`/`@Mutation()` method. You can tell Nest to execute interceptors, guards or filters for methods annotated with `@ResolveField()` by setting the `fieldResolverEnhancers` option in `GqlModuleOptions`.  Pass it a list of `'interceptors'`, `'guards'`, and/or `'filters'` as appropriate:
```

File: content/graphql/guards-interceptors.md [hunk 7]
```diff
@@ -85,3 +85,27 @@ async upvotePost(
...
+
+```typescript
+GraphQLModule.forRoot({
+  fieldResolverEnhancers: ['interceptors']
+}),
+```
+
+> **Warning** Enabling interceptors for field resolvers can cause performance issues when you are returning lots of records and your field resolver is executed thousands of times. For this reason, when you enable `fieldResolverEnhancers`, we advise you to skip execution of interceptors that are not strictly necessary for your field resolvers. You can do this using the following helper function:
```
Comment (johnbiundo) [inline content/graphql/guards-interceptors.md:99]:
```suggestion
> **Warning** Enabling enhancers for field resolvers can cause performance issues when you are returning lots of records and your field resolver is executed thousands of times. For this reason, when you enable `fieldResolverEnhancers`, we advise you to skip execution of enhancers that are not strictly necessary for your field resolvers. You can do this using the following helper function:
```

**Comment by gempain** [comment]
> Oh, I've missed one thing! Note that with the `fieldResolverEnhancers` you can enable interceptors, guards, and filters (not only interceptors). Example:
> 
> ```ts
> fieldResolverEnhancers: ['interceptors', 'guards']
> ```

I was aware but didn't know how to put it, and haven't tested with guards or filters, so I did not feel confident in writing about it. I forgot to mention it.

@kamilmysliwiec thanks for helping out improving the PR
@johnbiundo thanks for taking on suggesting change, very appreciated

File: content/graphql/guards-interceptors.md [hunk 8]
```diff
@@ -85,3 +85,27 @@ async upvotePost(
 ```
 
 > info **Hint** In the above example, we have assumed that the `user` object is assigned to the context of your GraphQL application.
+
+#### Execute enhancers at the field/method level
```
Comment (kamilmysliwiec) [inline content/graphql/guards-interceptors.md:89]: `/method` sounds confusing here. Query/mutation handlers are methods as well

**Comment by kamilmysliwiec** [comment]
Thanks @johnbiundo @gempain.
I left one minor comment

**Comment by johnbiundo** [comment]
@kamilmysliwiec Shall we just say this?
#### Execute enhancers at the field level

**Comment by gempain** [comment]
@johnbiundo the problem is that enhancers aren't documented anywhere, people don't know what they are. In a real world scenario (at least mine) you're looking for how to execute interceptors at the query/mutation handler level, and that's what this PR is about. I guess the docs should include something about what field enhancers are, but that's out of scope of this PR IMHO.

**Comment by kamilmysliwiec** [comment]
Thank you!

**Pull request merged**